internal/jobcontainers/logon.go

```diff
@@ -0,0 +1,68 @@
+package jobcontainers
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/Microsoft/hcsshim/internal/winapi"
+	"github.com/pkg/errors"
+	"golang.org/x/sys/windows"
+)
+
+// Takes in a DOMAIN\Username or just Username and will return a token
+// for the account if successful.
+func processToken(user string) (windows.Token, error) {
+	var (
+		domain   string
+		userName string
+		token    windows.Token
+	)
+
+	split := strings.Split(user, "\\")
+	if len(split) == 2 {
+		domain = split[0]
+		userName = split[1]
+	} else if len(split) == 1 {
+		userName = split[0]
+	} else {
+		return 0, fmt.Errorf("invalid user string `%s`", user)
+	}
+
+	// If empty just let it inherit the token from whatever is used to launch it (containerd-shim etc).
+	if user == "" {
+		return openCurrentProcessToken()
+	}
+
+	logonType := winapi.LOGON32_LOGON_INTERACTIVE
+	// User asking to run as a local system account (NETWORK SERVICE, LOCAL SERVICE, SYSTEM)
+	if domain == "NT AUTHORITY" {
+		logonType = winapi.LOGON32_LOGON_SERVICE
+	}
+
+	if err := winapi.LogonUser(
+		windows.StringToUTF16Ptr(userName),
+		windows.StringToUTF16Ptr(domain),
+		nil,
+		logonType,
+		winapi.LOGON32_PROVIDER_DEFAULT,
+		&token,
+	); err != nil {
+		// If we failed to grab a token for one of these accounts then just return the current processes token like we do for the
+		// no user was provided case. This handles the fallback case of there not being a user on the host that's actually named after one of the
+		// container virtual accounts. Normal Windows container images are still useable for job containers so the default usernames we'll get passed
+		// will still be ContainerUser and ContainerAdministrator, so we need to handle this edgecase.
+		if user == "ContainerUser" || user == "ContainerAdministrator" {
```

Comment:
Is there a specific "not found" error we can gate this on?

Reply:
Yea possibly, you mean like the below?
```go
if err == Win32UserNotFound && (user == "CotnainerUser" || user == "ContainerAdministrator" {}
```

Reply:
(Not the actual name of the error 😆, just clarifying what you meant)

Reply:
Yes, that's what I meant. Though I guess my first question should be "what's the impact if we just remove this?".

Reply:
As in this conditional? If so, just not being able to use the regular windows base images without supplying a user to run as or having a user on the host with no password and the same name as one of these two container virtual accounts. Seeming as how I'd imagine we'd regularly fail to find those two users, we'd just fail to start the container.

Reply:
Will this only affect use of the base images, or also propagate into derived images (assuming the user does not override the user in their Dockerfile)?

Reply:
Would be both if they never change the user to run as.

Reply:
That sounds quite impactful then, we would expect the majority of existing images to not run as a privileged container unless the user were overridden at runtime. However, maybe that's a better behavior anyways. If a user expects their container image to run as a standard user because "ContainerUser" was set, causing it to instead run as SYSTEM is a huge change to expectations. We should probably get input from sig-windows folks on this.

Reply:
The more I think about this, the more I feel we should just fail instead of falling back. We just need to make it clear to the user that they have to specify a real user to run as for a privileged container.

Reply:
Talked to Mark and he said let's discuss in the 1PM meeting with Jay and folks.